halpybot/packages/utils/spansh.py

```diff
@@ -0,0 +1,150 @@
+import aiohttp
+from loguru import logger
+from halpybot import config
+from . import shorten, web_get
+from ..exceptions import SpanshNoResponse, SpanshBadResponse, SpanshResponseTimedOut
+from ..models import Context, Points
+
+
+def sanitize_system_name(name: str) -> str:
+    """
+    Makes an effort to sanitize system input to help prevent SSRF or other web vulnerabilities
+    """
+    # Whitelist allowed characters (alphanumeric, hyphen, space)
+    allowed_chars = f"- {string.ascii_letters}{string.digits}+.'*()/#’,"
+    # Remove any disallowed characters from the system name
+    sanitized_name = "".join(c for c in name if c in allowed_chars)
+    return sanitized_name.strip()
+
+
+async def spansh_get_routes(
+    ctx: Context,
+    points: Points,
+    jobs: List[str],
+) -> None:
+    """
+    Receives calculated Normal and Neutron Jump Counts from spansh.co.uk
+
+    Args:
+        ctx (Context): PYDLE Context
+        points (Points): A pair of EDSM valid point locations and names, with a jump range
+        jobs (list): A list of two Spansh processing jobs
+
+    Returns:
+        None
+
+    Raises:
+        aiohttp.ClientError: Error occurred during HTTP Request
+        SpanshNoResponse: spansh did not respond in time.
+        SpanshBadResponse: Spansh returned an unprocessable response
+        SpanshResponseTimedOut: Spansh took too long to calculate a route
+    """
+    spansh_loop_timeout = config.spansh.calculations_timeout
+    job_results = []
+    for job in jobs:
+        while True:
+            try:  # Receive current job status
+                responses = await web_get(f"{config.spansh.results_endpoint}/{job}")
+            except aiohttp.ClientError as ex:
+                logger.exception(
+                    "spansh did not respond while trying to receive the calculation results"
+                )
+                raise SpanshNoResponse from ex
+            try:
+                if responses["status"] == "ok":
+                    # Spansh has finised calculations for this job, add all inidvidual jump counts together
+                    jumps = sum(
+                        entry["jumps"] for entry in responses["result"]["system_jumps"]
+                    )
+                    job_results.append(jumps)
+                    # Processing of this job has finished, start processing next job or respond with results
+                    break
+            except KeyError as keyerr:
+                logger.warning("Spansh returned an unprocessable response")
+                logger.warning(responses)
+                raise SpanshBadResponse from keyerr
+            if spansh_loop_timeout <= 0:
+                logger.exception("spansh took too long to calculate a route")
+                raise SpanshResponseTimedOut
+            # Wait 1 second to not send too many requests while waiting for spansh and update timeout counter
+            spansh_loop_timeout -= 1
+            await asyncio.sleep(1)
```

Comment:
this might be enough for *this* call, but doesn't protect againt concurrent calls